httpx/models.py

```diff
@@ -613,7 +620,16 @@ def prepare(self) -> None:
         has_connection = "connection" in self.headers
 
         if not has_host:
-            auto_headers.append((b"host", self.url.authority.encode("ascii")))
+            url = self.url
+            if url.userinfo:
+                url = url.copy_with_urlparse(userinfo=None)
```

Comment:
I'd rather we use the existing `.copy_with(username=None, password=None)`, instead of adding a new method.

I *do also* think we should also open a seperate issue to ensure that we support the *full* range of URL keyword arguments on `copy_with()`, so that eg. we support `.copy_with(userinfo=...)`, `copy_with(host=...)`, and other variants that `rfc3986` doesn't support directly.

Reply:
> I'd rather we use the existing `.copy_with(username=None, password=None)`, instead of adding a new method.
`URIReference` copy_with doesn't allow us to use `username` and `password` arguments. we can only use these: - scheme, authority, path, query, fragment

Reply:
Good point, yeah. Okay in that case I'd suggest that we've got a neccessary precursor to this pull request...

A *seperate pull request* that adds support for using `username, `password`, `hostname`, and `port` in our `copy_with(...)` method. There's an example of how we'd do that [in the Starlette codebase, here](https://github.com/encode/starlette/blob/6fa0038c699d2ac06f240d438263a4ae94e8996d/starlette/datastructures.py#L98-L119). (Tho replace "netloc" with "authority" in our case.)

Once we've got that in, then we can tackle this PR without having to add a new method to the API, which we don't really want to do.

Does that make sense / seem reasonable?

Thanks for digging through all this!

Reply:
Hi, @tomchristie I've pushed a new PR. After merge I will pursue from here